src/SonarScanner.MSBuild.PreProcessor/AnalysisConfigGenerator.cs

```diff
@@ -94,50 +103,61 @@ public static AnalysisConfig GenerateFile(ProcessedArgs localSettings,
-
-        // Ensure it isn't possible to write sensitive data to the config file
-        if (!property.ContainsSensitiveData())
-        {
-            properties.Add(new(id, value));
-        }
-    }
-
     // See https://sonarsource.atlassian.net/browse/SCAN4NET-29
-    // This method is a hack and should be removed when we properly support excluding coverage files in the scanner-engine.
-    // Instead, it should be replaced at the call site by "localSettings.CmdLineProperties.GetAllProperties()".
+    // This method is a hack and should be removed when we properly support excluding coverage files in the scanner-engine (https://sonarsource.atlassian.net/browse/SCANENGINE-18).
     // The idea is that we are manually adding the coverage paths to the exclusions, so that they do not appear on the analysis.
-    private static List<Property> GetCommandLineProperties(ProcessedArgs localSettings)
+    private static void HandleCoverageExclusions(AnalysisConfig config, ProcessedArgs localSettings, IDictionary<string, string> serverProperties)
     {
-        HashSet<string> coveragePropertyNames =
-        [
-            "sonar.cs.vscoveragexml.reportsPaths",
-            "sonar.cs.dotcover.reportsPaths",
-            "sonar.cs.opencover.reportsPaths",
-        ];
-        var allProperties = localSettings.CmdLineProperties.GetAllProperties().ToList();
-        var coveragePaths = string.Join(",", allProperties.Where(x => coveragePropertyNames.Contains(x.Id)).Select(x => x.Value));
-        if (!localSettings.ScanAllAnalysis      // if scanAll analysis is disabled, we will not pick up the coverage files anyways
-            || coveragePaths.Length == 0)       // if there are no coverage files, there is nothing to exclude
+        var localProperties = localSettings.AllProperties().ToList();
+        var localCoveragePaths = string.Join(",", localProperties.Where(x => CoveragePropertyNames.Contains(x.Id)).Select(x => x.Value));
+        var serverCoveragePaths = string.Join(",", serverProperties.Where(x => CoveragePropertyNames.Contains(x.Key)).Select(x => x.Value));
+        if (!localSettings.ScanAllAnalysis                                  // If scanAll analysis is disabled, we will not pick up the coverage files anyways
+            || localCoveragePaths.Length + serverCoveragePaths.Length == 0) // If there are no coverage files, there is nothing to exclude
         {
-            return allProperties;
+            return;
         }
+        var localExclusions = localSettings.GetSetting(SonarExclusions, string.Empty);
+        var serverExclusions = serverProperties.ContainsKey(SonarExclusions) ? serverProperties[SonarExclusions] : string.Empty;
+        AddCoverageExclusions(localCoveragePaths.Length > 0 ? localCoveragePaths : serverCoveragePaths);
 
-        if (allProperties.Find(x => x.Id == "sonar.exclusions") is { } exclusionsProperty)
+        void AddCoverageExclusions(string coveragePaths)
```

Comment:
What we are doing here is a bit of a war-crime.
I would leave a comment like this, for posterity:

```
        // The server exclusions are not passed to the scanner-cli. Instead, they are fetched from the server by the scanner-engine.
        // To prevent the coverage files of appearing in the UI, we need to override the local exclusions + coverage paths.
```